lib/test_summary_buildkite_plugin/tap.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+module TestSummaryBuildkitePlugin
+  # Parses most of the TAP protocol, assuming the inputs are sane.
+  #
+  # The specification is at https://testanything.org. This parses both
+  # version 12 and version 13.
+  #
+  # Notable omissions:
+  #
+  #     * Test numbering and the planned number of tests are ignored.
+  #     * "Bail out!" is ignored.
+  #
+  # Disclaimer:
+  #
+  #     This works about as well as you'd expect a hand-rolled parser made of
+  #     regular expressions to work. Use at your own risk, pull requests welcome.
+  #
+  # TODO: Use a proper grammar and parser rather than regexes.
+  class Tap
+    class Suite
+      attr_accessor :tests, :version
+
+      def initialize
+        self.version = 12
+        self.tests = []
+      end
+    end
+
+    Test = Struct.new(:passed, :description, :directive, :todo, :skipped, :diagnostic, :yaml,
+                      keyword_init: true)
+
+    class Parser
+      PATTERNS = {
+        plan: /^(?<start>\d+)\.\.(?<end>\d+)/,
+        test:
+          /^(?<not>not )?ok(?<number> \d+)?(?<description>[^#]*)(#\s*(?<directive>((?<todo>TODO)|(?<skip>SKIP))?.*))?/i,
+        comment: /^#(?<comment>.*)$/,
+        yaml_start: /^\s+---/,
+        yaml_end: /^\s+\.\.\./,
+        version: /^TAP version (?<version>\d+)/i
+      }.freeze
+
+      attr_reader :text
+      attr_reader :suite
+
+      def initialize(text)
+        @text = text
+        @suite = Suite.new
+        @current_diagnostic = []
+        @current_yaml = []
+        @in_yaml = []
```

Comment:
shan't this be initialized to false?

Reply:
maybe these three (current_diagnostic, current_yaml, in_yaml) should be thrown into a method with a name like `reset_parser_state!` to draw a line between the actual initialization and setting the parser state. Then you just call it here and in `save_previous_blocks` so that you always know you're resetting the parser state in a consistent way.